### PR TITLE
dynamically finding the backend preferred format for button images

### PR DESCRIPTION
**Pull request by fariza**
## PR Summary
The Tools used by Toolmanager assumed every backend understood `.png` files for the button images.
This is not the case, for example tk works with png in linux but not on mac

This pr fixes this by asking each backend for it's preferred image format

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/backend_bases.py [hunk 1]
```diff
@@ -3266,14 +3266,29 @@ def _remove_tool_cbk(self, event):
...
+
+        for fname in possible_images:
+            if os.path.isfile(fname):
+                return fname
+
+    def btn_image_extension(self):
+        """
+        Get the preferred image extension
+
+        Return
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: Returns.

File: lib/matplotlib/backend_bases.py [hunk 2]
```diff
@@ -3266,14 +3266,29 @@ def _remove_tool_cbk(self, event):
...
+        possible_images = [
+            image,
+            image + self.btn_image_extension(),
+            os.path.join(basedir, image) + self.btn_image_extension()]
+
+        for fname in possible_images:
+            if os.path.isfile(fname):
+                return fname
+
+    def btn_image_extension(self):
```
Comment (anntzer) [inline lib/matplotlib/backend_bases.py]: "preferred_image_extension"?

**Comment by anntzer** [comment]
FWIW recent tkinters should support png out of the box (https://stackoverflow.com/a/27601573/378080).

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (fariza) [reply lib/matplotlib/backend_bases.py]: prone to confusion?

Comment (fariza) [reply lib/matplotlib/backend_bases.py]: It is not actually a preference, it is matter of "does it work?"

Comment (anntzer) [reply lib/matplotlib/backend_bases.py]: supported_...

Comment (tacaswell) [reply lib/matplotlib/backend_bases.py]: could we get away with just having a class-level attribute of `_icon_extension` ?

Comment (anntzer) [reply lib/matplotlib/backend_bases.py]: probably would work too...

Comment (fariza) [reply lib/matplotlib/backend_bases.py]: That was my first idea, but I didn't like creating a property with a `raise AttributeError` in the base class.
But i guess you're right, it is alot easier and cleaner for the final implementation of the backend

Comment (anntzer) [reply lib/matplotlib/backend_bases.py]: Make it None on the base class which will make it raise a TypeError when fetching the attribute if not overridden?

**Comment by fariza** [comment]
@tacaswell  I changed it to a class attribute as suggested, I think it is cleaner this way

**Comment by anntzer** [comment]
Repro'ing my comment that got folded when the code was patched away:
> Make it None on the base class which will make it raise a TypeError when fetching the attribute if not overridden?

File: lib/matplotlib/backend_bases.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/backend_bases.py]: Or default to `'png'`?

**Comment by tacaswell** [comment]
I think it is ok to leave it as `'png'` as the default extension as well.

**Comment by fariza** [comment]
Done!, added a default format

**Comment by dstansby** [comment]
I think this will need a manual rebase against current `master` for circleCI to work.

**Pull request merged**